fairscale/nn/data_parallel/offload_ddp.py

```diff
@@ -0,0 +1,402 @@
+            will be used.
+
+        broadcast_buffers (bool, optional):
+            whether to sync all the model buffers at the beginning of a FW pass
+
+        shard_split_strategy(`SplitStrategy`):
+            how to shard the model. Can be iso-layer or iso-parameter size
+
+        offload_optimizer (bool):
+            move optimizer computations to the offload device (for instance CPU), which saves more memory but slows down
+    """
+
+    def __init__(
+        self,
+        model_cpu: nn.Sequential,  # hard pre-requisite for now, easier model slicing
+        optimizer: Type[torch.optim.Optimizer],
+        optimizer_params: Dict[str, Any],
+        world_size: int,
+        device: torch.device,
+        offload_device: torch.device = torch.device("cpu"),
+        process_group: Any = None,
+        broadcast_buffers: bool = True,
+        shard_split_strategy: SplitStrategy = SplitStrategy.MEMORY,
+        offload_optimizer: bool = False,
+    ):
+        super().__init__()
+
+        self.world_size = world_size
+        self.process_group = process_group if process_group is not None else torch.distributed.group.WORLD
+        self.rank = dist.get_rank(self.process_group)
+        self.global_rank = self.get_global_rank(self.process_group, self.rank)
+        self.backend = dist.get_backend(group=self.process_group)
+        self.device = device
+        self.offload_device = device
+
+        # Slice the model into roughly equivalent sequential shards
+        splits = _split(model_cpu, self.world_size, shard_split_strategy)
+
+        # Each rank either owns the slice, or temporarily helps processing it in a data parallel fashion
+        self.model_slices: List[nn.Module] = []
+
+        for i_slice, module_shard in enumerate(splits):
+            global_owner_rank = self.get_global_rank(self.process_group, i_slice)
+
+            # Add one dataparallel model handling this slice
+            self.model_slices.append(
+                ModelShard(
+                    cpu_model_shard=nn.Sequential(*module_shard),
+                    owner_rank=global_owner_rank,
+                    process_group=self.process_group,
+                    device=device,
+                    offload_device=offload_device,
+                    broadcast_bufers=broadcast_buffers,
+                    offload_optimizer=False,
+                )
+            )
+
+            # Use one normal optimizer per shard
+            if i_slice == self.rank:
+                self.optimizer = optimizer(nn.Sequential(*module_shard).parameters(), **optimizer_params)
+
+        # Expose a unified view of the slices
+        self.model = torch.nn.Sequential(*self.model_slices)
+        self.sync_ranks()
+
+    def forward(self, *inputs: Any, **kwargs: Any) -> Any:
+        # Slice per slice FW, sync in between
+        syncRanks = ShardSyncLayer.apply
+
+        for i, (p2, p1, n1, n2) in enumerate(
```

Comment:
this part is a bit tricky, the idea is to pre-load the next-next shard (and vice versa in the BW pass) in parallel from the current compute, instead of this blocking the compute wavefront